### PR TITLE
Fix and test the LU solver for FV

### PR DESCRIPTION
**Pull request by mwarusz**
This PR update the LU solver to support both DG and DGFV system

- DG: lower_bandwidth is Nq_v*nstate * eband - 1, (does not include itself) 
    - eband = 1 for inviscid, since nodal point at the face communicates to the overlaping point to its neighbor
       and other points only communicates to points in the same element
    - eband = 2 for visous

- FV: lower_bandwidth is nstate * (stencil_width + 1 + 1) - 1,  
   - since the reconstruction states depend on stencil_width points on each side, 
      and the flux depends on stencil_width + 1 points on each side
   - eband = (stencil_width + 2) for inviscid
   - eband = max{ (stencil_width + 2), 3} for viscid, 
   - since the viscous flux is computed by applying first order FD twice

Tests are added:

- The fully implicit solver with Jacobian free Newton Krylov method and LU precondition is add in test/Numerics/DGMethods/advection_diffusion/fvm_advection_diffusion_model_1dimex_bjfnks.jl
- Unit tests are added in test/Numerics/SystemSolvers/columnwiselu.jl
- Unit tests are added in test/Numerics/SystemSolvers/bandedsystem.jl

Some findings:

- For linear problem (fvm_advection_diffusion_model_1dimex_bjfnks.jl), since we have preconditioner
  - with constant reconstruction, it converges in 1 iteration to machine accuracy
  - with Linear reconstruction, turning off the limiter, it converges in 1 iteration  to machine accuracy
  - with Linear reconstruction, turning on the limiter, it converges in 5 iteration  to machine accuracy. The limiter makes it a nonlinear problems
   

- [x] Code follows the [style guidelines](https://clima.github.io/ClimateMachine.jl/latest/DevDocs/CodeStyle/) OR N/A.
- [x] Unit tests are included OR N/A.
- [x] Code is exercised in an integration test OR N/A.
- [x] Documentation has been added/updated OR N/A.

### CODE REVIEW

**Comment by mwarusz** [comment]
bors try

**Comment by jkozdon** [comment]
@mwarusz Do you want me to look at this before or after you've updated it with master?

**Comment by mwarusz** [comment]
Let me update it first. BTW: I think it fails CI only because I added a test for N=0 and we run CI with bounds checking so it picks up the out of bounds access in interface tendency boundary conditions.

**Comment by Zhengyu-Huang** [comment]
Nice work!   
Do we also need a user defined stencil_size here, the stencil size can be  1 3 5 and 
               1: first order FV,         3 diagonal matrix for convection 
               3: second order FV,   5 diagonal matrix for convection
               5: WENO5,                 7 diagonal matrix for convection
                                                   5 diagonal matrix for diffusion 
And the LU bandwidth should also depend on it.
@jkozdon

**Comment by jkozdon** [comment]
> Do we also need a user defined stencil_size here, the stencil size can be 1 3 5

Yeah we will probably need something like this, but let's wait until we have that functionality added and know what the interface will be

**Comment by Zhengyu-Huang** [comment]
I can help update this PR for FVConstant and FVLinear and update the driver

**Comment by Zhengyu-Huang** [comment]
bors try

**Review by Zhengyu-Huang — APPROVED** [review]
Hi Maciej,  I updated your PR, and the changes are documented in the comment. 
I approved the change, you can review it and decide when to merge it.

Thank you for your great efforts!

**Comment by Zhengyu-Huang** [comment]
bors try

File: src/Numerics/SystemSolvers/columnwise_lu_solver.jl [hunk 1]
```diff
@@ -52,9 +52,26 @@ elem_band(
 single_column(
     ::DGColumnBandedMatrix{D, P, NS, EH, EV, EB, SC},
 ) where {D, P, NS, EH, EV, EB, SC} = SC
+
+# DG: lower_bandwidth is Nq_v*nstate * eband - 1, (does not include itself) 
+# eband = 1 for inviscid, since nodal point at the face communicates to the overlaping point to its neighbor
+# and other points only communicates to points in the same element
```
Comment (valeriabarra) [inline src/Numerics/SystemSolvers/columnwise_lu_solver.jl:58]:
```suggestion
# and other points only communicate to points in the same element
```

File: src/Numerics/SystemSolvers/columnwise_lu_solver.jl [hunk 2]
```diff
@@ -52,9 +52,26 @@ elem_band(
...
+# eband = 2 for visous
+#
+# FV: lower_bandwidth is nstate * (stencil_width + 1 + 1) - 1,  
+# since the reconstruction states depend on stencil_width points on each side, 
+# and the flux depends on stencil_width + 1 points on each side
+# eband = (stencil_width + 2) for inviscid
+# eband = max{ (stencil_width + 2), 3} for viscid,
```
Comment (valeriabarra) [inline src/Numerics/SystemSolvers/columnwise_lu_solver.jl:65]:
```suggestion
# eband = max{ (stencil_width + 2), 3} for viscous, 
```
just for consistency with line 59 above

File: src/Numerics/SystemSolvers/columnwise_lu_solver.jl [hunk 3]
```diff
@@ -52,9 +52,26 @@ elem_band(
...
+#
+# FV: lower_bandwidth is nstate * (stencil_width + 1 + 1) - 1,  
+# since the reconstruction states depend on stencil_width points on each side, 
+# and the flux depends on stencil_width + 1 points on each side
+# eband = (stencil_width + 2) for inviscid
+# eband = max{ (stencil_width + 2), 3} for viscid, 
+# since the viscous flux is computed by applying first order FD twice
+#  
+# The lower_bandwidth ls formulated as Nq_v*nstate * eband - 1
+lower_bandwidth(N, nstate, eband) = (N + 1) * nstate * eband - 1
```
Comment (valeriabarra) [inline src/Numerics/SystemSolvers/columnwise_lu_solver.jl:69]:
```suggestion
# The lower_bandwidth ls formulated as Nq_v * nstate * eband - 1
Nq_v = N + 1
lower_bandwidth(N, nstate, eband) = Nq_v * nstate * eband - 1
```
I think this might just read a bit nicer with the comment above and might be a bit an easier refactoring if/when we implement overintegration where `Nq_v = N + 1 + qextra`, where `qextra` witll be the number of extra quadrature points

**Comment by valeriabarra** [comment]
This overall looks good to me. But my review of the tests was a bit shallow. If anyone wants to also review would be great. Otherwise, @mwarusz do you want to go ahead and merge this?

**Comment by mwarusz** [comment]
I made a few small changes including @valeriabarra  suggestions and this PR looks good to me. Thanks for helping with this @Zhengyu-Huang.
